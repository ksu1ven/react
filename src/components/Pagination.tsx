import React from 'react';
import { useRouter } from 'next/router';
import { updateQueryParams } from '../utils/helpFunctions';

interface Props {
  pageNumber: number;
  totalPages: number;
  paginationButtonsValue: number[];
}

function Pagination(props: Props) {
  const { pageNumber, totalPages, paginationButtonsValue } = props;

  const router = useRouter();

  function changePaginationButtonsValue() {
    if (
      totalPages - paginationButtonsValue[paginationButtonsValue.length - 1] <=
      1
    )
      return;
    const increaseNumber =
      totalPages - paginationButtonsValue[paginationButtonsValue.length - 1] <
      paginationButtonsValue.length
        ? 1
        : paginationButtonsValue.length;
    const newPageNumber =
      paginationButtonsValue[paginationButtonsValue.length - 1] +
      increaseNumber;
    router.push(
      '?' + updateQueryParams(router.query, 'page', newPageNumber.toString())
    );
  }

  return (
    <div className="flex justify-center gap-4 mb-10" data-testid="pagination">
      <button
        className="w-14 h-14 text-lime-700 text-4xl p-3 rounded-full text-white font-extrabold"
        onClick={() => {
          if (pageNumber <= 0) return;
          router.push(
            '?' + updateQueryParams(router.query, 'page', pageNumber.toString())
          );
        }}
      >
        {'<'}
      </button>
      {paginationButtonsValue.map((value, ind) => {
        return (
          totalPages > value && (
            <button
              data-testid={`page-${ind + 1}-button`}
              key={`button-${value}`}
              className={
                pageNumber + 1 === value
                  ? 'w-14 h-14 bg-lime-300 p-3 rounded-full text-white font-extrabold'
                  : 'w-14 h-14 bg-lime-700 p-3 rounded-full text-white font-extrabold'
              }
              onClick={() => {
                router.push(
                  '?' +
                    updateQueryParams(router.query, 'page', value.toString())
                );
              }}
            >
              {value}
            </button>
          )
        );
      })}

      {totalPages > paginationButtonsValue.length && (
        <button
          className="w-14 h-14 text-lime-700 text-4xl p-3 rounded-full text-white font-extrabold"
          onClick={() => {
            changePaginationButtonsValue();
          }}
        >
          ...
        </button>
      )}
      <button
        className={
          pageNumber + 1 === totalPages
            ? 'w-14 h-14 bg-lime-300 p-3 rounded-full text-white font-extrabold'
            : 'w-14 h-14 bg-lime-700 p-3 rounded-full text-white font-extrabold'
        }
        data-testid={'last-page-button'}
        onClick={() => {
          router.push(
            '?' + updateQueryParams(router.query, 'page', totalPages.toString())
          );
        }}
      >
        {totalPages}
      </button>
      <button
        className="w-14 h-14 text-lime-700 text-4xl p-3 rounded-full text-white font-extrabold"
        onClick={() => {
          if (pageNumber + 1 > totalPages - 1) return;
          router.push(
            '?' +
              updateQueryParams(
                router.query,
                'page',
                (pageNumber + 2).toString()
              )
          );
        }}
      >
        {'>'}
      </button>
    </div>
  );
}

export default Pagination;
