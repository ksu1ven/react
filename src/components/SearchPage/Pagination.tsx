interface Props {
  pageNumber: number;
  setPageNumber: React.Dispatch<React.SetStateAction<number>>;
  totalPages: number;
}

function Pagination(props: Props) {
  const { pageNumber, setPageNumber, totalPages } = props;

  return (
    <div className="flex justify-center gap-4 mb-10">
      <button className="w-14 h-14 text-lime-700 text-4xl p-3 rounded-full text-white font-extrabold">
        {'<'}
      </button>
      <button
        className={
          pageNumber === 0
            ? 'w-14 h-14 bg-lime-300 p-3 rounded-full text-white font-extrabold'
            : 'w-14 h-14 bg-lime-700 p-3 rounded-full text-white font-extrabold'
        }
        onClick={() => {
          setPageNumber(pageNumber > 3 ? pageNumber - 2 : 1);
        }}
      >
        {pageNumber > 3 ? pageNumber - 3 : 1}
      </button>
      {totalPages > 2 && (
        <button
          className={
            pageNumber === 1
              ? 'w-14 h-14 bg-lime-300 p-3 rounded-full text-white font-extrabold'
              : 'w-14 h-14 bg-lime-700 p-3 rounded-full text-white font-extrabold'
          }
          onClick={() => {}}
        >
          {pageNumber > 3 ? pageNumber - 2 : 2}
        </button>
      )}
      {totalPages > 3 && (
        <button
          className={
            pageNumber === 2
              ? 'w-14 h-14 bg-lime-300 p-3 rounded-full text-white font-extrabold'
              : 'w-14 h-14 bg-lime-700 p-3 rounded-full text-white font-extrabold'
          }
          onClick={() => {
            setPageNumber(pageNumber > 2 ? pageNumber : 3);
          }}
        >
          {pageNumber > 3 ? pageNumber - 1 : 3}
        </button>
      )}
      {totalPages > 3 && (
        <button
          className="w-14 h-14 text-lime-700 text-4xl p-3 rounded-full text-white font-extrabold"
          onClick={() => {
            setPageNumber(pageNumber + 1);
          }}
        >
          ...
        </button>
      )}
      <button className="w-14 h-14 bg-lime-700 p-3 rounded-full text-white font-extrabold">
        {totalPages}
      </button>
      <button className="w-14 h-14 text-lime-700 text-4xl p-3 rounded-full text-white font-extrabold">
        {'>'}
      </button>
    </div>
  );
}

export default Pagination;
