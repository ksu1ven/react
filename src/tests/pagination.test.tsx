import React from 'react';
import { describe, expect, it, vi } from 'vitest';
import '@testing-library/jest-dom';
import { fireEvent, render, screen, cleanup } from '@testing-library/react';
import mockRouter from 'next-router-mock';
import { MemoryRouterProvider } from 'next-router-mock/MemoryRouterProvider';
import Pagination from '../components/Pagination';

vi.mock('next/router', () => vi.importActual('next-router-mock'));

describe('Tests for the Pagination component', () => {
  it('The component updates URL query parameter when page changes', async () => {
    mockRouter.push('/');
    render(
      <Pagination
        pageNumber={(Number(mockRouter.query.page) - 1) | 0}
        totalPages={200}
        paginationButtonsValue={[1, 2, 3]}
      />,
      {
        wrapper: MemoryRouterProvider,
      }
    );
    expect(screen.getByTestId('pagination')).toMatchSnapshot();
    expect(await screen.findByTestId('pagination')).toBeInTheDocument();
    fireEvent.click(screen.getByText('3'));

    expect(mockRouter.asPath).toBe('/?page=3');
    cleanup();

    render(
      <Pagination
        pageNumber={(Number(mockRouter.query.page) - 1) | 0}
        totalPages={200}
        paginationButtonsValue={[3, 4, 5]}
      />,
      {
        wrapper: MemoryRouterProvider,
      }
    );
    expect(await screen.findByTestId('page-1-button')).toHaveClass(
      'w-14 h-14 bg-lime-300 p-3 rounded-full text-white font-extrabold'
    );
    fireEvent.click(await screen.findByText('>'));
    expect(mockRouter.asPath).toBe('/?page=4');

    cleanup();

    render(
      <Pagination
        pageNumber={(Number(mockRouter.query.page) - 1) | 0}
        totalPages={200}
        paginationButtonsValue={[3, 4, 5]}
      />,
      {
        wrapper: MemoryRouterProvider,
      }
    );
    fireEvent.click(await screen.findByText('...'));
    expect(mockRouter.asPath).toBe('/?page=8');

    cleanup();

    render(
      <Pagination
        pageNumber={(Number(mockRouter.query.page) - 1) | 0}
        totalPages={200}
        paginationButtonsValue={[8, 9, 10]}
      />,
      {
        wrapper: MemoryRouterProvider,
      }
    );
    expect(screen.getByTestId('page-1-button')).toHaveTextContent('8');
    expect(screen.getByTestId('page-2-button')).toHaveTextContent('9');
    expect(screen.getByTestId('page-3-button')).toHaveTextContent('10');
    fireEvent.click(await screen.findByTestId('last-page-button'));
    expect(mockRouter.asPath).toBe('/?page=200');
  });
});
