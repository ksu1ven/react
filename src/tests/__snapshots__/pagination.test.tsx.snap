// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for the Pagination component > The component updates URL query parameter when page changes 1`] = `
<div
  class="flex justify-center gap-4 mb-10"
  data-testid="pagination"
>
  <button
    class="w-14 h-14 text-lime-700 text-4xl p-3 rounded-full text-white font-extrabold"
  >
    &lt;
  </button>
  <button
    class="w-14 h-14 bg-lime-700 p-3 rounded-full text-white font-extrabold"
    data-testid="page-1-button"
  >
    1
  </button>
  <button
    class="w-14 h-14 bg-lime-700 p-3 rounded-full text-white font-extrabold"
    data-testid="page-2-button"
  >
    2
  </button>
  <button
    class="w-14 h-14 bg-lime-300 p-3 rounded-full text-white font-extrabold"
    data-testid="page-3-button"
  >
    3
  </button>
  <button
    class="w-14 h-14 text-lime-700 text-4xl p-3 rounded-full text-white font-extrabold"
  >
    ...
  </button>
  <button
    class="w-14 h-14 bg-lime-700 p-3 rounded-full text-white font-extrabold"
    data-testid="last-page-button"
  >
    20
  </button>
  <button
    class="w-14 h-14 text-lime-700 text-4xl p-3 rounded-full text-white font-extrabold"
  >
    &gt;
  </button>
</div>
`;
